src/components/App/App.js

```diff
@@ -23,308 +23,336 @@ import newsApi from "../../utils/NewsApi";
+     setHasResults(true);
+    }
+   })
+   .catch((err) => {
+    console.log(err);
+    setHasError(true);
+   })
+   .finally(() => {
+    setIsLoading(false);
+   });
+ }
+
+ function handleLogOut() {
+  setLoggedIn(false);
+  localStorage.removeItem("jwt");
+  history.push("/");
+ }
+
+ function handleLogin() {
+  setHasError(false);
+  setLoggedIn(true);
+  setIsSignInOpen(false);
+ }
+
+ function handleRegister() {
+  setHasError(false);
+  setIsSignUpOpen(false);
+  setIsSuccessfulPopupOpen(true);
+ }
+
+ function handleSignInClick() {
+  setHasError(false);
+  setIsSignInOpen(true);
+  setIsSignUpOpen(false);
+  setIsSuccessfulPopupOpen(false);
+ }
+
+ function handleSignUpClick() {
+  setHasError(false);
+  setIsSignUpOpen(true);
+  setIsSignInOpen(false);
+ }
+
+ function handleRegisterSubmit(email, password, name) {
+  auth
+   .register(email, password, name)
+   .then((res) => {
+    if (res) {
+     setIsRegistered(true);
+     handleRegister();
+    } else {
+     setIsRegistered(false);
+     setHasError(true);
+    }
+   })
+   .catch((err) => {
+    console.log(`This email is unavailable: ${err.message}`);
+    setHasError(true);
+   });
+ }
+
+ function handleLoginSubmit(email, password) {
+  auth
+   .login(email, password)
+   .then((data) => {
+    if (data.token) {
+     localStorage.setItem("jwt", data.token);
+     setToken(data.token);
+     handleLogin();
+     history.push("/");
+    }
+   })
+   .catch((err) => {
+    console.log(`Incorrect email or password: ${err.message}`);
+    setHasError(true);
+   });
+ }
+
+ function closeAllPopups() {
+  setIsSignInOpen(false);
+  setIsSignUpOpen(false);
+  setIsSuccessfulPopupOpen(false);
+ }
+
+ return (
+  <CurrentUserContext.Provider value={currentUser}>
+   <div className="app">
+    <Header
+     loggedIn={loggedIn}
+     currentUser={currentUser}
+     setLoggedIn={setLoggedIn}
+     onSignInClick={handleSignInClick}
+     setIsNewsCardListOpen={setIsNewsCardListOpen}
+     setSearchKeyword={setSearchKeyword}
+     onSavedArticlesPage={onSavedArticlesPage}
+     onLogOut={handleLogOut}
+    />
+    <Switch>
+     <Route exact path="/">
+      <SearchForm
+       onSearch={handleSearchSubmit}
+       searchKeyword={searchKeyword}
+       setSearchKeyword={setSearchKeyword}
+       setIsNewsCardListOpen={setIsNewsCardListOpen}
+      />
+      {hasResults && isNewsCardListOpen && (
+       <NewsCardList
+        onSavedArticlesPage={onSavedArticlesPage}
+        loggedIn={loggedIn}
+        cards={cards}
+        savedArticles={savedArticles}
+        onSaveArticleClick={handleSaveArticle}
+        onRemoveArticleClick={handleRemoveArticle}
+        showCards={showCards}
+        setShowCards={setShowCards}
+        onSignInClick={handleSignInClick}
+       />
+      )}
+
+      {isLoading && <Preloader />}
+      {!hasResults && !isLoading && isNewsCardListOpen && (
+       <NotFoundResults hasError={hasError} />
+      )}
+      <About />
+     </Route>
+
```

Comment:
**Could be improved**


If it’s interesting for you here is a way to make a special wrapper for any `popup` with cross button, `Escape` and `overlay` listeners inside. You should create component `Popup`, store it in `Popup.js`. 

```jsx
import { useEffect } from "react";

const Popup = ({ isOpen, name, onClose, children }) => {
// here is `useEffect` for the `Escape` listener
  useEffect(() => {
    // with this we prevent adding the listener if the popup is not opened
    if (!isOpen) return;
// we should define the handler inside `useEffect`, so that it wouldn’t lose the reference to be able to remove it
    const closeByEscape = (e) => {
      if (e.key === 'Escape') {
        onClose();
      }
    }

    document.addEventListener('keydown', closeByEscape)
    // don’t forget to remove the listener in the `clean-up` function
    return () => document.removeEventListener('keydown', closeByEscape)
// here we watch `isOpen` to add the listener only when it’s opened
}, [isOpen, onClose])

// here is the overlay handler
  const handleOverlay = (e) => {
    if (e.target === e.currentTarget) {
        onClose();
    }
  }

// then we add the main wrapper with class `popup` and `popup_opened`
  return (
    <div
      className={`popup ${isOpen ? "popup_opened" : ""} popup_type_${name}`}
      onClick={handleOverlay}
    >
    {/* the container for the contents */}
      <div className='popup__container'>
        {/* here will be anything you add as `children`
*/}
        {children}
        {/* add the close button */}
        <button
          className='popup__close'
          type='button'
          onClick={onClose}
        />
      </div>
    </div>
  );
};

export default Popup;


```

And now you can use `Popup` for any popup in the project:  ` ImagePopup`, ` InfoToolTip` and `PopupWithForm`.

```jsx


function PopupWithForm({isOpen, name, onClose, ...props}) {
  return (
    <Popup isOpen={isOpen} name={name} onClose={onClose}>
        <h2 className='popup__title'>{props.title}</h2>


```

You can make also `Input` for inputs and `Form` for any form in the project.